File: packages/core/src/gpu/buffer-pool.ts
import type { BufferEntry } from "../types.js";

/** Create a GPU uniform buffer from an ArrayBuffer, with mappedAtCreation. */
export function createUniformBuffer(device: GPUDevice, data: ArrayBuffer): GPUBuffer {
  const size = Math.max(Math.ceil(data.byteLength / 4) * 4, 4);
  const buffer = device.createBuffer({
    size,
    usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    mappedAtCreation: true,
  });
  new Uint8Array(buffer.getMappedRange()).set(new Uint8Array(data));
  buffer.unmap();
  return buffer;
}

/**
 * GPU buffer pool for reusing temporary compute buffers.
 * Reduces allocation churn during inference by recycling buffers
 * that match the requested size (rounded up to alignment).
 */
export class BufferPool {
  private device: GPUDevice;
  private pools = new Map<number, BufferEntry[]>();
  private bufferToEntry = new Map<GPUBuffer, BufferEntry>();
  private alignment: number;

  constructor(device: GPUDevice, alignment = 256) {
    this.device = device;
    this.alignment = alignment;
  }

  /** Round size up to the nearest alignment boundary */
  private alignSize(size: number): number {
    return Math.ceil(size / this.alignment) * this.alignment;
  }

  /**
   * Acquire a buffer of at least `size` bytes with the given usage flags.
   * Returns a pooled buffer if one is available, otherwise creates a new one.
   */
  acquire(size: number, usage: GPUBufferUsageFlags): GPUBuffer {
    const aligned = this.alignSize(size);
    const pool = this.pools.get(usage);

    if (pool) {
      for (const entry of pool) {
        if (!entry.inUse && entry.size >= aligned) {
          entry.inUse = true;
          return entry.buffer;
        }
      }
    }

    const buffer = this.device.createBuffer({ size: aligned, usage });
    const entry: BufferEntry = { buffer, size: aligned, inUse: true };
    this.bufferToEntry.set(buffer, entry);

    if (!pool) {
      this.pools.set(usage, [entry]);
    } else {
      pool.push(entry);
    }

    return buffer;
  }

  /** Release a buffer back to the pool for reuse. */
  release(buffer: GPUBuffer): void {
    const entry = this.bufferToEntry.get(buffer);
    if (entry) {
      entry.inUse = false;
    }
  }

  /** Destroy all pooled buffers and clear the pool. */
  destroy(): void {
    for (const pool of this.pools.values()) {
      for (const entry of pool) {
        entry.buffer.destroy();
      }
    }
    this.pools.clear();
    this.bufferToEntry.clear();
  }
}

File: packages/rust/crates/oxbitnet/src/gpu/buffer_pool.rs
use std::sync::Arc;

use wgpu::BufferUsages;

/// A shareable reference to a GPU buffer.
pub type GpuBuf = Arc<wgpu::Buffer>;

/// GPU buffer pool / factory.
///
/// For MVP, this simply creates fresh buffers. A future optimization
/// can add size-bucketed reuse (like the TS BufferPool).
pub struct BufferPool {
    device: Arc<wgpu::Device>,
    alignment: u64,
}

impl BufferPool {
    pub fn new(device: Arc<wgpu::Device>, alignment: u64) -> Self {
        Self { device, alignment }
    }

    fn align_size(&self, size: u64) -> u64 {
        ((size + self.alignment - 1) / self.alignment) * self.alignment
    }

    /// Create a buffer of at least `size` bytes with the given usage flags.
    pub fn acquire(&self, size: u64, usage: BufferUsages) -> GpuBuf {
        let aligned = self.align_size(size.max(4));
        Arc::new(self.device.create_buffer(&wgpu::BufferDescriptor {
            label: None,
            size: aligned,
            usage,
            mapped_at_creation: false,
        }))
    }

    /// Release a buffer (currently a no-op; buffer is destroyed when Arc drops).
    pub fn release(&self, _buffer: GpuBuf) {
        // Drop — destroyed when last Arc ref is gone.
    }
}
